### PR TITLE
feat(zkevm-circuits): enable type 1 and 2 as an input data

### PR DESCRIPTION
**Pull request by dongchangYoo**
There are 2 changes as below:

feat(zkevm-circuits): enable type 1 and 2 as a input data
- fix bug on handling access list
- ensure that the address in access list and tx's 'from/to'
  are only written once to state db in circuit input builder.
  If not, `Lookup Rw` error occurs

refac(integration-test): refactoring integration test
- The content of the gendata_output file is guaranteed.
  (If txs are not in a block, it will be dead.)
- reduced the amount of reusable code.
- simplify the setup process logs.
- add test case for type 2 tx with access list

### CODE REVIEW

**Comment by dongchangYoo** [comment]
Due to a minor oversight, a few forced pushes were performed. As there are no further tasks remaining, I kindly request a review.

**Comment by fakedev9999** [comment]
@dongchangYoo for the third commit message, let's fix "refactoring" to "refactor", following our convention to use imperative mood.

File: zkevm-circuits/src/super_circuit.rs [hunk 1]
```diff
@@ -972,7 +1018,7 @@ pub(crate) mod super_circuit_tests {
 
     #[test]
     fn serial_test_super_circuit_2tx_2max_tx() {
-        let block = block_2tx();
+        let block = block_type2_1tx();
         const MAX_TXS: usize = tx_idx!(2);
```
Comment (fakedev9999) [inline zkevm-circuits/src/super_circuit.rs]:
```suggestion
        const MAX_TXS: usize = tx_idx!(1);
```

File: integration-tests/src/bin/gen_blockchain_data.rs [hunk 2]
```diff
@@ -156,240 +49,162 @@ async fn main() {
...
-        wallets[0].clone(),
-        contracts.get("Greeter").expect("contract not found"),
-        U256::from(42),
-    )
-    .await;
-    blocks.insert(
-        "Deploy Greeter".to_string(),
-        receipt.block_number.unwrap().as_u64(),
-    );
+    // Deploy "Greter" contract
```
Comment (fakedev9999) [inline integration-tests/src/bin/gen_blockchain_data.rs]:
```suggestion
    // Deploy "Greeter" contract
```

File: integration-tests/src/bin/gen_blockchain_data.rs [hunk 3]
```diff
@@ -156,240 +49,162 @@ async fn main() {
...
+    let contract_name = "Greeter";
+    let target_contarct = contracts.get(contract_name).expect("contarct not found");
```
Comment (fakedev9999) [inline integration-tests/src/bin/gen_blockchain_data.rs]:
```suggestion
    let target_contract = contracts.get(contract_name).expect("contract not found");
```

File: integration-tests/src/bin/gen_blockchain_data.rs [hunk 4]
```diff
@@ -156,240 +49,162 @@ async fn main() {
...
-            .get("OpenZeppelinERC20TestToken")
-            .expect("contract not found"),
-        wallets[0].address(),
-    )
-    .await;
-    blocks.insert(
-        "Deploy OpenZeppelinERC20TestToken".to_string(),
-        receipt.block_number.unwrap().as_u64(),
-    );
+    // Deploy "ERC20" contract
```
Comment (fakedev9999) [inline integration-tests/src/bin/gen_blockchain_data.rs]:
```suggestion
    // Deploy "OpenZeppelinERC20TestToken" contract
```

File: integration-tests/src/bin/gen_blockchain_data.rs [hunk 5]
```diff
@@ -156,240 +49,162 @@ async fn main() {
...
-        contracts.get("Greeter").expect("contract not found"),
-        U256::from(42),
-    )
-    .await;
-    blocks.insert(
-        "Deploy Greeter".to_string(),
-        receipt.block_number.unwrap().as_u64(),
-    );
+    // Deploy "Greter" contract
+    let contract_name = "Greeter";
```
Comment (TomTaehoonKim) [inline integration-tests/src/bin/gen_blockchain_data.rs:58]:
```suggestion
    let contract_name = "Greeter";
```

File: integration-tests/src/bin/gen_blockchain_data.rs [hunk 6]
```diff
@@ -156,240 +49,162 @@ async fn main() {
...
-        wallets[0].address(),
-    )
-    .await;
-    blocks.insert(
-        "Deploy OpenZeppelinERC20TestToken".to_string(),
-        receipt.block_number.unwrap().as_u64(),
-    );
+    // Deploy "ERC20" contract
+    let contract_name = "OpenZeppelinERC20TestToken";
+    let target_contarct = contracts.get(contract_name).expect("contarct not found");
```
Comment (fakedev9999) [inline integration-tests/src/bin/gen_blockchain_data.rs]:
```suggestion
    let target_contract = contracts.get(contract_name).expect("contract not found");
```

File: integration-tests/src/bin/gen_blockchain_data.rs [hunk 7]
```diff
@@ -156,240 +49,162 @@ async fn main() {
         "Transfer 0".to_string(),
         receipt.block_number.unwrap().as_u64(),
     );
+    info!("- done(height: {:?})", receipt.block_number);
```
Comment (fakedev9999) [inline integration-tests/src/bin/gen_blockchain_data.rs]:
```suggestion
    info!("- Done(height: {:?})", receipt.block_number);
```
Can you apply all the changes to Done? I suggest this change because we use "Doing" capitalized.

File: integration-tests/src/scenario_utils.rs [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+            from_i,
+            to_i,
+            amount,
+            expected_result,
+        }
+    }
+}
+
+/// Build and return erc20 tranfer transaction
+/// TODO(dongchangYoo) Currently this function does not support that an account sends tx twice in a
```
Comment (fakedev9999) [inline integration-tests/src/scenario_utils.rs]:
```suggestion
/// TODO(dongchangYoo): Currently this function does not support sending multiple txs from an account in a
```

File: integration-tests/src/scenario_utils.rs [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+    solc::Solc,
+    types::TransactionRequest,
+    utils::WEI_IN_ETHER,
+};
+use log::{error, info};
+use std::{collections::HashMap, fs::File, path::Path, sync::Arc, thread::sleep, time::Duration};
+
+use crate::{get_provider, get_wallet, CompiledContract, CONTRACTS, CONTRACTS_PATH};
+
+/// Number of transaction in each scenario
```
Comment (fakedev9999) [inline integration-tests/src/scenario_utils.rs]:
```suggestion
/// Number of transactions in each scenario
```

File: integration-tests/src/scenario_utils.rs [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+            Err(err) => {
+                error!("Geth not available: {:?}", err);
+                sleep(Duration::from_millis(500));
+            }
+        }
+    }
+    prov
+}
+
+/// init wallets to be used in each scenario
```
Comment (fakedev9999) [inline integration-tests/src/scenario_utils.rs]:
```suggestion
/// Init wallets to be used in each scenario
```

File: integration-tests/src/scenario_utils.rs [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+        )
+        .expect("cannot serialize json into file");
+
+        contracts.insert(name.to_string(), compiled_contract);
+    }
+
+    contracts
+}
+
+/// Depoloy contract
```
Comment (fakedev9999) [inline integration-tests/src/scenario_utils.rs]:
```suggestion
/// Deploy contract
```

File: integration-tests/src/scenario_utils.rs [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+        .unwrap_or_else(|_| panic!("cannot deploy: {:?}", compiled.name))
+        .confirmations(0usize)
+        .send_with_receipt()
+        .await
+        .unwrap_or_else(|_| panic!("cannot deploy: {:?}", compiled.name));
+
+    (receipt.block_number.unwrap().as_u64(), contract)
+}
+
+/// An `sender` account sends ETH to multiple accounts
```
Comment (fakedev9999) [inline integration-tests/src/scenario_utils.rs]:
```suggestion
/// A `sender` account sends ETH to multiple accounts
```

File: integration-tests/src/scenario_utils.rs [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+                "transfer",
+                (wallets[trans_ctx.to_i].address(), trans_ctx.amount),
+            )
+            .expect("cannot construct ERC20 transfer call");
+
+        // Set gas to avoid `eth_estimateGas` call
+        let call = call.gas(100_000);
+
+        // Convert tx to legacy tx if tx_type equals to 0
+        // TODO(dongchangYoo) use constant for tx_type after implementing various tx types
```
Comment (fakedev9999) [inline integration-tests/src/scenario_utils.rs]:
```suggestion
        // TODO(dongchangYoo): use constant for tx_type after implementing various tx types
```

File: integration-tests/src/scenario_utils.rs [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+            amount,
+            expected_result,
+        }
+    }
+}
+
+/// Build and return erc20 tranfer transaction
+/// TODO(dongchangYoo) Currently this function does not support that an account sends tx twice in a
+/// function.
+pub async fn exchange_token(
```
Comment (fakedev9999) [inline integration-tests/src/scenario_utils.rs]:
```suggestion
pub async fn transfer_erc20_token(
```
Or is the "exchange" a formal term to use in a situation of ERC-20 token transfers?

File: .gitignore [hunk 15]
```diff
@@ -5,4 +5,6 @@
 .idea
 *.log
 *.json
-*.sh
\ No newline at end of file
+*.sh
+*.h
+*.so
```
Comment (chokobole) [inline .gitignore]: Why did you add this?

File: zkevm-circuits/src/super_circuit.rs [hunk 16]
```diff
@@ -869,6 +869,52 @@ pub(crate) mod super_circuit_tests {
         block
     }
 
+    pub(crate) fn block_type2_1tx() -> GethData {
```
Comment (chokobole) [inline zkevm-circuits/src/super_circuit.rs]: For me, it's better to name `dynamic fee tx` rather than `type2 tx`.

```suggestion
    pub(crate) fn block_1dynamic_fee_tx() -> GethData {
```

File: integration-tests/src/scenario_utils.rs [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+
+        pending_txs.push(
+            wallets[trans_ctx.from_i]
+                .send_transaction(tx, None)
+                .await
+                .expect("cannot send token"),
+        )
+    }
+
+    sleep(Duration::from_secs(5));
```
Comment (chokobole) [inline integration-tests/src/scenario_utils.rs]: Please remove this!

File: integration-tests/src/scenario_utils.rs [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+
+        // Convert tx to legacy tx if tx_type equals to 0
+        // TODO(dongchangYoo) use constant for tx_type after implementing various tx types
+        let tx = match tx_type {
+            0 => {
+                let call = call.legacy();
+                call.tx
+            }
+            _ => call.tx,
+        };
```
Comment (chokobole) [inline integration-tests/src/scenario_utils.rs]:
```suggestion
        let tx = if tx_type == 0 { 
            let call = call.legacy();
            call.tx
        } else {
            call.tx
        };
```

File: integration-tests/src/scenario_utils.rs [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+
+        pending_txs.push(
+            wallets[trans_ctx.from_i]
+                .send_transaction(tx, None)
+                .await
+                .expect("cannot send token"),
```
Comment (chokobole) [inline integration-tests/src/scenario_utils.rs]:
```suggestion
                .expect("cannot transfer token"),
```

File: integration-tests/src/scenario_utils.rs [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+/// TransferContext
+pub struct TransferContext {
+    /// Index of `from` wallet
+    pub from_i: usize,
+    /// Index of `to` wallet
+    pub to_i: usize,
+    /// Amount to transfer
+    pub amount: U256,
+    /// Expected result
+    pub expected_result: bool,
```
Comment (chokobole) [inline integration-tests/src/scenario_utils.rs]: Let's simplify the names

```suggestion
    /// Index of `from` wallet
    pub from: usize,
    /// Index of `to` wallet
    pub to: usize,
    /// Amount to transfer
    pub amount: U256,
    /// Expected result
    pub expected: bool,
```

File: integration-tests/src/scenario_utils.rs [hunk 21]
```diff
@@ -0,0 +1,236 @@
...
+        pending_txs.push(
+            wallets[0]
+                .send_transaction(tx, None)
+                .await
+                .expect("cannot send tx"),
+        );
+        nonce = nonce.checked_add(U256::one()).unwrap();
+    }
+
+    sleep(Duration::from_secs(5));
```
Comment (chokobole) [inline integration-tests/src/scenario_utils.rs]: Please remove this!

File: integration-tests/src/scenario_utils.rs [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+    block_num = 0;
+    for tx in pending_txs.into_iter() {
+        let receipt = tx.await.expect("cannot confirm tx").unwrap();
+        if block_num == 0 {
+            block_num = receipt.block_number.unwrap().as_u64();
+        } else if block_num != receipt.block_number.unwrap().as_u64() {
+            panic!("The txs are not in a block")
+        }
+    }
+    block_num
```
Comment (chokobole) [inline integration-tests/src/scenario_utils.rs]: Please extract this to a function and do not repeat this.

File: integration-tests/src/scenario_utils.rs (hunk 19 above)
Comment (chokobole) [reply integration-tests/src/scenario_utils.rs]: Above, you used "cannot send tx" for the error message. Please unify the error message.

File: integration-tests/src/scenario_utils.rs [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+    args: T,
+) -> (u64, Contract<M>)
+where
+    M: Middleware,
+    T: Tokenize,
+{
+    let factory = ContractFactory::new(compiled.abi.clone(), compiled.bin.clone(), wallet.clone());
+    let (contract, receipt) = factory
+        .deploy(args)
+        .unwrap_or_else(|_| panic!("cannot deploy: {:?}", compiled.name))
```
Comment (chokobole) [inline integration-tests/src/scenario_utils.rs]:
```suggestion
        .unwrap_or_else(|_| panic!("cannot construct contract creation tx: {:?}", compiled.name))
```

**Comment by chokobole** [comment]
How about changing commit titles like below?

- `build(.gitignore): change gitignore` -> `chore: change gitignore`
- `feat(zkevm-circuits): enable type 1 and 2 as a input data` -> `feat(zkevm-circuits): enable type 1 and 2 as an input data`
- `refac(integration-test): refactoring integration test` -> `test(integration-test): improve integration test`

File: integration-tests/src/scenario_utils.rs [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+    T: Tokenize,
+{
+    let factory = ContractFactory::new(compiled.abi.clone(), compiled.bin.clone(), wallet.clone());
+    let (contract, receipt) = factory
+        .deploy(args)
+        .unwrap_or_else(|_| panic!("cannot construct contract creation tx: {:?}", compiled.name))
+        .confirmations(0usize)
+        .send_with_receipt()
+        .await
+        .unwrap_or_else(|_| panic!("cannot construct contract creation tx: {:?}", compiled.name));
```
Comment (chokobole) [inline integration-tests/src/scenario_utils.rs]:
```suggestion
        .unwrap_or_else(|_| panic!("cannot deploy tx: {:?}", compiled.name));
```

File: integration-tests/src/bin/gen_blockchain_data.rs [hunk 25]
```diff
@@ -156,240 +49,162 @@ async fn main() {
...
+        &prov,
+        contract_address,
+        contract_abi,
+        &wallets,
+        transfer_map,
+        2,
+    )
+    .await;
+    blocks.insert(case_name.to_string(), block_num);
+    info!("- done(height: {:?})", block_num);
```
Comment (fakedev9999) [inline integration-tests/src/bin/gen_blockchain_data.rs]:
```suggestion
    info!("- Done(height: {:?})", block_num);
```

File: integration-tests/src/scenario_utils.rs [hunk 26]
```diff
@@ -0,0 +1,223 @@
...
+            to: to_i,
+            amount,
+            expected: expected_result,
+        }
+    }
+}
+
+/// Build and return erc20 transfer transaction
+/// TODO(dongchangYoo): Currently this function does not support sending multiple txs from an
+/// account in a
```
Comment (fakedev9999) [inline integration-tests/src/scenario_utils.rs]:
```suggestion
/// account.
```
Sorry, I gave you an incomplete suggestion before.

**Review by lightscale-luke — APPROVED** [review]
LGTM

**Review by chokobole — APPROVED** [review]
LGTM

**Review by fakedev9999 — APPROVED** [review]
Good job!! Looks good to go~
